lib/web/apiserver.go

```diff
@@ -278,54 +284,78 @@ func (m *Handler) Close() error {
 	return m.auth.Close()
 }
 
-type oidcConnector struct {
-	ID      string `json:"id"`
-	Display string `json:"display"`
-}
-
-type webSettings struct {
-	Auth struct {
-		OIDCConnectors []oidcConnector `json:"oidc_connectors"`
-		U2FAppID       string          `json:"u2f_appid"`
-	} `json:"auth"`
-}
-
 func (m *Handler) getUserStatus(w http.ResponseWriter, r *http.Request, _ httprouter.Params, c *SessionContext) (interface{}, error) {
 	return ok(), nil
 }
 
-func (m *Handler) getSettings(w http.ResponseWriter, r *http.Request) (interface{}, error) {
-	settings := &webSettings{}
-	connectors, err := m.cfg.ProxyClient.GetOIDCConnectors(false)
+func buildAuthenticationSettings(authClient auth.ClientI) (*client.PingResponse, error) {
+	as := &client.AuthenticationSettings{}
+
+	cap, err := authClient.GetClusterAuthPreference()
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	for _, connector := range connectors {
-		fmt.Printf("%v\n", connectors)
-		settings.Auth.OIDCConnectors = append(settings.Auth.OIDCConnectors, oidcConnector{
-			ID:      connector.GetName(),
-			Display: connector.GetDisplay(),
-		})
+	as.Type = cap.GetType()
+	as.SecondFactor = cap.GetSecondFactor()
+
+	if cap.GetSecondFactor() == "u2f" {
+		universalSecondFactor, err := authClient.GetUniversalSecondFactor()
+		if err != nil {
+			if !trace.IsNotFound(err) {
+				log.Infof("Unable to get U2F Settings: %v", err)
+				goto oidc
```

Comment:
can you split into two functions? `goto` makes it harder to read than it should probably be